provider/sidecar.go

```diff
@@ -123,42 +124,44 @@ func (provider *Sidecar) loadSidecarConfig() error {
 }
 
 func (provider *Sidecar) sidecarWatcher() error {
-	client := &http.Client{Timeout: 0}
+	//set timeout to be just a bot more than connection refresh interval
+	client := &http.Client{Timeout: (provider.RefreshConn + 3) * time.Second}
 	resp, err := client.Get(provider.Endpoint + "/watch")
 	if err != nil {
 		return err
 	}
-	err = catalog.DecodeStream(resp.Body, provider.callbackLoader)
-	if err != nil {
-		return err
-	}
-	for { //check for open http connection, try to re-init if it fails
-		if resp.Close {
-			resp, err = client.Get(provider.Endpoint + "/watch")
-			if err != nil {
-				return err
-			}
-			err = catalog.DecodeStream(resp.Body, provider.callbackLoader)
-			if err != nil {
-				return err
-			}
+	go catalog.DecodeStream(resp.Body, provider.callbackLoader)
+	log.Infof("Using %s Sidecar connection refresh interval", provider.RefreshConn)
+	provider.recycleConn(resp, client)
+	return nil
+}
+
+func (provider *Sidecar) recycleConn(resp *http.Response, client *http.Client) {
+	var err error
+	for { //use refresh interval to occasionally reconnect to Sidecar in case the stream connection is lost
+		time.Sleep(provider.RefreshConn * time.Second)
+		resp, err = client.Get(provider.Endpoint + "/watch")
```

Comment:
The idea here was to reuse the http.Response variable ("resp") on each loop and let the GC handle the cleanup.  As a test to validate this I added a log message to show the number of goroutines using [runtime.NumGoroutine()](https://golang.org/pkg/runtime/#NumGoroutine).  After its run for 20 minutes, running the loop every 10 seconds (rovider.RefreshConn set to 10 seconds) , the number of goroutines has held steady at 25 the entire time.  This seemed like a valid test to me, but maybe you know of a caveat I overlooked or another test I can try to verify the number of goroutines isn't getting out of hand?

The difficulty with ensuring the connection is up is the stream isn't ever going to close the connection on the server side.  So json.Decode will just wait for the next object; it will not return an error if it (unknowingly) loses an existing connection.  Checking the http.response.close works until you actually lose the connection.  Then a nil error occurs which is a pretty ugly way to health check.